src/main/java/com/multithreading/TextProcessor.java

```diff
@@ -1,38 +1,54 @@
 package com.multithreading;
 
+import java.util.concurrent.atomic.AtomicInteger;
+
 class Capitalize implements Runnable {
     private final char[] chars;
     private final int index;
+    private final AtomicInteger numTasksLeft;
 
-    public Capitalize(char[] chars, int index) {
+    public Capitalize(char[] chars, int index, AtomicInteger numTasksLeft) {
         this.chars = chars;
         this.index = index;
+        this.numTasksLeft = numTasksLeft;
     }
 
     @Override
     public void run() {
         synchronized (chars) {
             chars[index] = Character.toUpperCase(chars[index]);
         }
+        synchronized (numTasksLeft) {
+            numTasksLeft.decrementAndGet();
+            numTasksLeft.notifyAll();
+        }
     }
 }
 
 public class TextProcessor {
 
-    public String process(String input) throws InterruptedException {
-        Thread[] threads = new Thread[input.length()];
+    private final ThreadPool threadPool;
 
+    public TextProcessor(int numThreads) {
+        this.threadPool = new ThreadPool(numThreads);
```

Comment:
Consider reading about inversion of dependencies. In this case instead of passing numThreads, you may want to pass the threadpool itself:
```
main(){
   pool = new ThreadPool(NUM_THREADS);
   new Processor(pool);
}
```

That will also help with testing.